### PR TITLE
feat: ADL nested stack support

### PR DESCRIPTION
**Pull request by mndeveci**
#### Which issue(s) does this change fix?
N/A

#### Why is this change necessary?
In order to support Auto Dependency Layer creation for Nested Stacks.

#### How does it address the issue?
Previous implementation was only creating dependency layers for root stack. With this change, we are iterating over all stacks and create auto dependency layers for functions.

#### What side effects does this change have?
N/A

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/bootstrap/nested_stack/nested_stack_manager.py [hunk 1]
```diff
@@ -105,18 +105,21 @@ def generate_auto_dependency_layer_stack(self) -> Dict:
             LOG.debug("No function has been added for auto dependency layer creation")
             return template
 
-        nested_template_location = os.path.join(self._build_dir, "nested_template.yaml")
-        move_template(self._stack_location, nested_template_location, self._nested_stack_builder.build_as_dict())
+        nested_template_location = str(self._get_template_folder().joinpath("nested_template.yaml"))
```
Comment (mndeveci) [inline samcli/lib/bootstrap/nested_stack/nested_stack_manager.py:108]: Change `nested_template.yaml` to something related to ADL

File: samcli/lib/bootstrap/nested_stack/nested_stack_builder.py [hunk 2]
```diff
@@ -43,11 +45,13 @@ def add_function(
     @staticmethod
     def get_layer_logical_id(function_logical_id: str) -> str:
         function_logical_id_hash = str_checksum(function_logical_id)
+        function_logical_id = NestedStackBuilder._get_logical_id_compliant_str(function_logical_id)
         return f"{function_logical_id[:48]}{function_logical_id_hash[:8]}DepLayer"
 
     @staticmethod
     def get_layer_name(stack_name: str, function_logical_id: str) -> str:
         function_logical_id_hash = str_checksum(function_logical_id)
+        function_logical_id = NestedStackBuilder._get_logical_id_compliant_str(function_logical_id)
```
Comment (mildaniel) [inline samcli/lib/bootstrap/nested_stack/nested_stack_builder.py:54]: nit: Can we rename this variable here and above to something more specific to describe that it's been mutated?

**Review by mildaniel — APPROVED** [review]

**Review by mingkun2020 — APPROVED** [review]
LGTM!

File: samcli/lib/bootstrap/nested_stack/nested_stack_builder.py (hunk 2 above)
Comment (mndeveci) [reply samcli/lib/bootstrap/nested_stack/nested_stack_builder.py]: Good suggestion, addressed in the latest commit 👍🏼

**Pull request merged**